### PR TITLE
Credentials API on SDK/Agent

### PR DESCRIPTION
**Pull request by mnzaki**
This adds `CredentialKeeper` and `CredentialIssuer` classes, and instances of them on the `sdk` and `agent`.
They can be used as `sdk.credentials.get`, `create`/`issue`, `query`, `delete`, and also similarly on the `agent`
On the agent, the credentials scope is limited to those whose subject is the agent (need to also add support for checking for issuer!)

This supersedes and includes changes from #93 but refactored to use `ObjectKeeper`

There's a circular dependency on https://github.com/jolocom/sdk-storage-typeorm/pull/19

### CODE REVIEW

File: src/credentials.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+export interface CredentialQuery {
+  id?: string
+  issuer?: string
+  subject?: string
+  type?: string[]
+  claim?: ClaimEntry
+}
+
+export class CredentialKeeper
```
[the remark this answers is not in the document]
Comment (mnzaki) [reply src/credentials.ts:111]: This already supports getting all credentials: `sdk.credentials.find()` to get all credentials of all agents, or `agent.credentials.find()` to get all credentials for an agent. Basically if you don't supply a query then it lists all of them. Is that sufficient?

Also btw this is not a breaking change, the old access patterns through `sdk.storage` of course will still work, this is just a nicer wrapper around them

File: src/credentials.ts [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  }
+
+  /**
+   * Creates and signs a Credential
+   *
+   * @param credParams - credential attributes
+   * @returns SignedCredential instance
+   * @category Credential Management
+   */
+  async create<T extends BaseMetadata>(credParams: ISignedCredCreationArgs<T>) {
```
[the remark this answers is not in the document]
Comment (mnzaki) [reply src/credentials.ts:184]: Yes good point, I haven't updated the rest of the code to use the new API!

File: src/interactionManager/credentialOfferFlow.ts [hunk 3]
```diff
@@ -161,4 +162,12 @@ export class CredentialOfferFlow extends Flow<
...
     })
   }
+
+  public getOfferDisplay() {
+    return this.state.offerSummary.map((oc, idx) => {
+      const claim = this.state.issued[idx]?.claim
+      const credType = new CredentialType(oc.type, oc.credential || {} as CredentialDefinition)
+      return credType.display(claim)
+    })
+  }
```
[the remark this answers is not in the document]
Comment (mnzaki) [reply src/interactionManager/credentialOfferFlow.ts:172]: They do slightly different things, because one of them (`CredentialKeeper.display`) retrieves the credential definition from storage, while this one (`CredentialOfferFlow.getOfferDisplay`) uses the definition that's in the offer token in this interaction.
It can be improved perhaps by making `CredentialKeeper.display` take an optional `credDefinition`, or otherwise look it up from storage. Then we can use that method here on the `offerSummary` entries

File: src/credentials.ts [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    return {
+      ...offer,
+      credential: {
+        ...credentialDefaults,
+        ...this.def,
+        ...offer.credential,
+      },
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (mnzaki) [reply src/credentials.ts:101]: Agreed, this needs to be removed. The functionality is specific to constructing a credential offer, and currently that code is actually in `agent.credOfferToken`, not in the `CredentialOfferFlow`. This is a whole other can of worms (the API for constructing tokens for different phases of an interaction) that I hope we can address as part of an upcoming SDK architecture improvement round

I think `CredentialType` should remain public though. The name could perhaps be improved, but its purpose is to literally represents a 'kind' or 'type' of credential. Instances of this class should hold metadata like the name of this type of credential, data schema, display information, etc.
Basically it's meant as a holder of information about those pesky [Verifiable Credential types](https://w3c.github.io/vc-data-model/#types)

File: src/interactionManager/credentialOfferFlow.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mnzaki) [reply src/interactionManager/credentialOfferFlow.ts]: and also one is sync and one is async :sweat_smile:
